### ISSUE TITLE
Task: Explore Evennia to Python3 conversion

### ISSUE THREAD

**Issue by Griatch** [evennia/evennia#1481]
#### Description of requested feature:

This task is for exploring just how far we can get with Evennia + Python3. 

> I'm not sure how suitable this actually is for hacktoberfest (it's unclear how big of an effort it is). It's not suitable for a Python-beginner but most likely you don't need that much Evennia-specific knowledge so it might be interesting to explore.

Currently, Evennia is solely a Python2 program. It's clear that Python3 is the future. What has held us up on this have historically been our dependencies. Django has supported py3 since several versions at this point. The remaining big dependency-block for us is Twisted, and their Py3 conversion is, by their own account, [96% complete](https://twistedmatrix.com/trac/milestone/Python-3.x). And [here is the Evennia project page tracking py3](https://github.com/evennia/evennia/projects/5). It is currently unclear if those remaining issues are blockers for us or not. 

With Django un-supporting Python2 completely in version 2.0 and Python2 itself reaching end-of-upstream-support in 2020, I don't think there is a good reason to spend effort on a Python2 + Python3 hybrid of Evennia. Only reason for going for a hybrid would be if it turns out this is somehow easier or helps the process. 

#### Process:

In a separate branch, try to run Evennia on python3, to see just how far one can get with the current state of third-party libraries.

1. Install python3
2. Set up virtualenv with py3: `virtualenv -p /path/to/python3/executable pyenv`
3. Install Evennia into this virtualenv - py3 versions of third-party packages should be used automatically. 
4. See what works and what not, and fix if possible.

#### Expected result of task:

Ideally, an Evennia running under Python3. Otherwise a report on what is missing in the third-party libraries for that to be possible (we won't patch third-party packages). 

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:

This should be done against `develop` branch.

**Comment by chainsol** [comment]
So I spent some time this morning working on this. I hit a couple roadblocks, used 2to3 to convert most of the minor differences, but:

- Evennia stores the settings values in ServerConfig.value, which is a TextField. It has to be changed to a BinaryField, otherwise you'll end up with an error about converting utf-8.
- We use a utility function, to_str, a lot, and in Python 3, when you run `str(b"whatever")`, you end up with a literal `b"whatever"` string. This gets especially odd when we're using that for database access. This method actually might not be needed - I overrode it to just return obj if it's an instance of str.
- The AMP connection seems to loop - possibly I did something wrong converting to Python 3.
- Assuming we fix these issues, twisted.web.server seems to not be serving the static files for some reason.

Other than that, Evennia launched, you could visit the site, but since AMP didn't work, you couldn't actually connect to Evennia over any method.

**Comment by Griatch** [comment]
For people looking at this, and to avoid someone reinventing the wheel, user MudCastle (don't know their github handle) has continued the work of chainsol towards py3 compliance and has gotten really far.  So don't start any work on this before checking in IRC first (so as to not reinvent the wheel).

**Comment by glyph** [comment]
Should this issue be closed?  All the CI appears to be on Python 3.7 :)

**Comment by Griatch** [comment]
@glyph I suppose it can be closed yes. Traditionally we keep `develop`-branch issues open until they merge into `master`, but this one is a little confusing and outdated now.
